packages/orm/src/EntityManager.ts

```diff
@@ -522,12 +556,25 @@ export class EntityManager {
     const joinRowTodos = sortJoinRows(this.__data.joinRows);
 
     if (Object.keys(entityTodos).length > 0 || Object.keys(joinRowTodos).length > 0) {
-      await this.knex.transaction(async (tx) => {
-        await flushEntities(this.knex, tx, entityTodos);
-        await flushJoinTables(this.knex, tx, joinRowTodos);
-        await tx.commit();
-      });
+      const alreadyInTxn = "commit" in this.knex;
+
+      if (!alreadyInTxn) {
+        await this.knex.transaction(async (knex) => {
+          await flushEntities(knex, entityTodos);
+          await flushJoinTables(knex, joinRowTodos);
+          // When using `.transaction` with a lambda, we don't explicitly call commit
+          // await knex.commit();
+        });
+      } else {
+        await flushEntities(this.knex, entityTodos);
+        await flushJoinTables(this.knex, joinRowTodos);
+        // Defer to the caller to commit the transaction
+      }
+
+      // TODO: This is really "after flush" if we're being called from a transaction that
+      // is going to make multiple `em.flush()` calls?
```

Comment:
The wrinkle here is that a valid api usage would be:

```typescript
em.transaction(async () => {
  em.find...
  em.create...
  // sends sql to db, doesn't commit
  await em.flush();

  em.find...
  em.create...
  // send more sql to db
  await em.flush()
}); // lambda is over, now txn commit happens
```

In the current impl, our `afterCommit` hooks would be ran for each `em.flush` call.

I'm pretty sure this is fine for this PR / to fix our current blueprint bug, but is something to think about for handling better/differently/potentially at some point.

Reply:
maybe just change the name to `afterFlush` instead of `afterCommit`?